### PR TITLE
feature: add support for request/response interceptors, support CORS with credentials

### PR DESCRIPTION
**Pull request by tommytroen**
* Move CORS related functionality into interceptor
* Support CORS requests with credentials by allowing credentials and specific origin

### CODE REVIEW

**Comment by jp7677** [comment]
Hi @tommytroen Could you please also push the file containing the `CorsInterceptor` ;)

**Comment by tommytroen** [comment]
> Hi @tommytroen Could you please also push the file containing the `CorsInterceptor` ;)

Oops :) Pushed now!

**Comment by tommytroen** [comment]
> Hi @tommytroen I've build your branch and started a container with it. When the open-id configuration is fetched, I'm seeing correct response headers, but the response body appears to be empty. I wont rule out that I did something wrong (essentially `git pull https://github.com/navikt/mock-oauth2-server feature/interceptors` on top of `master` and `./gradlew jibDockerBuild --image mock-oauth2-server:0.5-SNAPSHOT1`) considering the tests are succeeding but may be you see something obvious.

Hmm.. Strange. I ran the container from the branch and requested the config via browser (not with origin) - at least then it seem to work as before. Did you test with the origin header set?

**Comment by jp7677** [comment]
> > Hi @tommytroen I've build your branch and started a container with it. When the open-id configuration is fetched, I'm seeing correct response headers, but the response body appears to be empty. I wont rule out that I did something wrong (essentially `git pull https://github.com/navikt/mock-oauth2-server feature/interceptors` on top of `master` and `./gradlew jibDockerBuild --image mock-oauth2-server:0.5-SNAPSHOT1`) considering the tests are succeeding but may be you see something obvious.
> 
> Hmm.. Strange. I ran the container from the branch and requested the config via browser (not with origin) - at least then it seem to work as before. Did you test with the origin header set?

Yes, I have the origin header set. When not doing this, thus requesting directly in the browser, then I'm also seeing the response body.

File: src/main/kotlin/no/nav/security/mock/oauth2/http/OAuth2HttpRouter.kt [hunk 1]
```diff
@@ -56,40 +73,60 @@ interface Route : RequestHandler {
...
+    private fun RequestHandler.with(interceptors: MutableList<Interceptor>): RequestHandler {
+        return { request ->
+            val filteredRequest = interceptors.filterIsInstance<RequestInterceptor>().fold(request) { next, interceptor ->
+                interceptor.intercept(next)
+            }
+            val res = this.invoke(filteredRequest)
+            interceptors.filterIsInstance<ResponseInterceptor>().fold(res) { next, interceptor ->
+                interceptor.intercept(request, next)
+            }
+        }
```
Comment (tommytroen) [inline src/main/kotlin/no/nav/security/mock/oauth2/http/OAuth2HttpRouter.kt:115]: @jp7677 Found the bug. Forgot to return the intercepted response here 🤦 I'll create some more tests, fix the bug and commit! Thanks for testing!

**Comment by tommytroen** [comment]
@jp7677 Seems the bug is a bit more intricate so I'll probably need some more time to debug.

**Comment by jp7677** [comment]
> @jp7677 So I think I've located the bug... It appeared to be in the `CorsInterceptor` itself in relation to using Netty as the underlying server. By mistake I copied the request headers to a new header builder for the response, this included the header `Content-Length: 0` which by Netty was interpreted as an empty response body. Not sure how to add a test that catches these kind of bugs.. Anyways, it seems to be working on my machine now :) Can you test again?

Hi @tommytroen , thanks a lot for diving into the rabbit hole. this looks indeed like a tricky one. 
Good news, all good here, works as intended. The concept with the filter looks also very cool and a lot cleaner than bolting it into the responses like I did.

**Comment by jp7677** [comment]
> @jp7677 That sounds really strange - the refresh token is supposed to be an opaque string containing whatever. Do you mean that keycloak requires the refresh token to be a JWT? Maybe I havent understood the use case right - could you elaborate a bit more on what you send, what you receive and what errors keycloak-js is reporting?

The refresh token being an opaque string is also my understanding, but seems that unfortunately keycloak enhanced the specs for themself. See here https://github.com/keycloak/keycloak/blob/main/adapters/oidc/js/src/keycloak.js#L952 where it wants to decode the refresh token like an id-token and here https://github.com/keycloak/keycloak/blob/main/adapters/oidc/js/src/keycloak.js#L765 where it validates the nonce element in the refresh token.

**Comment by jp7677** [comment]
@tommytroen Do you still plan to move forward with the CORS changes?
Additionally, should I move the posting about the refresh token into a separate issue? I guess that is more suitable to continue thoughts about that topic. Sry for not doing that in the first place.

**Comment by tommytroen** [comment]
> @tommytroen Do you still plan to move forward with the CORS changes? Additionally, should I move the posting about the refresh token into a separate issue? I guess that is more suitable to continue thoughts about that topic. Sry for not doing that in the first place.

@jp7677 Sounds like a plan, separate issue would be good. I'll merge this PR then 🚢

**Pull request merged**